storage/client/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java

```diff
@@ -144,23 +161,61 @@ public static BlockBlobClientBuilder builder() {
      * @return
      *      A reactive response containing the information of the uploaded block blob.
      */
-    public Mono<Void> upload(Flux<ByteBuffer> data, long length, BlobHTTPHeaders headers,
+    public Mono<Response<BlockBlobItem>> upload(Flux<ByteBuffer> data, long length, BlobHTTPHeaders headers,
             Metadata metadata, BlobAccessConditions accessConditions, Context context) {
         return blockBlobAsyncRawClient
             .upload(data.map(Unpooled::wrappedBuffer), length, headers, metadata, accessConditions, context)
-            .then();
+            .map(rb -> new SimpleResponse<>(rb, new BlockBlobItem(rb.deserializedHeaders())));
     }
 
     public Mono<Void> uploadFromFile(String filePath) {
         return this.uploadFromFile(filePath, null, null, null, null);
     }
 
     public Mono<Void> uploadFromFile(String filePath, BlobHTTPHeaders headers, Metadata metadata,
-            BlobAccessConditions accessConditions, Context context) {
-        //TODO make this method smart
-        return this.blockBlobAsyncRawClient
-            .upload(ByteBufFlux.fromPath(Paths.get(filePath)), new File(filePath).length())
-            .then();
+                                     BlobAccessConditions accessConditions, Context context) {
+        AsynchronousFileChannel channel;
+        try {
+            channel = AsynchronousFileChannel.open(Paths.get(filePath), StandardOpenOption.READ);
```

Comment:
Similar to the download to file method, we should:

- Use Mono.using()
- support progres reporting
- support the possibility of doing a single putBlob

Reply:
Same question for progress reporting. For a single putBlob, I believe it's the upload() method. Otherwise addressed.